sourcegraph/sourcegraph
# web: add react tracer and instrument the blob page

## Context

This PR introduces a couple of components and hooks to facilitate our React code instrumentation. It is the first stab into the React instrumentation that gives us data about component mount timings relative to the page load event and other key components of instrumented pages.

Usage example:

```tsx
const Traced = () => (
  <TraceSpanProvider name="RepositoryFileTreePage">
      <RepositoryFileTreePage {...props} />,
  </TraceSpanProvider>
)
```

`TraceSpanProvider` creates the OpenTelemetry tracing span on the first component render call, and it ends the span automatically on the component mount event. Does nothing and returns `children` if OpenTelemetry tracing is disabled.

Closes https://github.com/sourcegraph/sourcegraph/issues/39282
Closes https://github.com/sourcegraph/sourcegraph/issues/40467

## Example

[Honeycomb trace link](https://ui.honeycomb.io/sourcegraph/environments/valery-testing/datasets/web-app/result/ASQJGDYnfQc/trace/BcHbCfT9HMb?fields[]=c_name&fields[]=c_http.url&span=d6bc2068ca0e414b)

<img width="1206" alt="Screen Shot 2022-08-17 at 17 05 11" src="https://user-images.githubusercontent.com/3846380/185080013-11a105a6-6b25-4a73-be9b-5c9732585be6.png">

## Why use React context instead of OpenTelemetry context manager?

The proposed implementation [uses](https://github.com/sourcegraph/sourcegraph/blob/7562b2cfcaf579c53283f40b104922ac27d106f9/client/observability-client/src/react/TraceSpanProvider/TraceSpanProvider.tsx) React context to correctly connect spans created by nested components.
 
[The OpenTelemetry context](https://opentelemetry.io/docs/instrumentation/js/api/context/) is immutable and [can only be passed down with a callback](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/995#issuecomment-1138367723). If there's no way to wrap the function execution into a parent span via callback, we need to implement another sharing mechanism like a store. This issue is raised in the OpenTelemetry repo and currently does not have a recommended solution.

## Test plan

1. `ENABLE_OPEN_TELEMETRY=true sg start web-standalone`
2. Open the browser console and ensure that debug level output is enabled.
3. Navigate between to the Blob page. 
4. See navigation events logged into the console with nested events created by React components.

<img width="444" alt="Screen Shot 2022-08-18 at 22 46 14" src="https://user-images.githubusercontent.com/3846380/185424375-10a11e82-4fd7-46e9-845c-bc04f890f560.png">

The web application should function as usual with disabled OpenTelemetry.

1. `sg start web-standalone`
2. Ensure that the Blob page works as expected.

## App preview:

- [Web](https://sg-web-vb-react-trace-provider.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-ctcudjjbxa.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
